### PR TITLE
Feature/eab 885 transform eab2 payload for epms

### PR DESCRIPTION
**Pull request by gardnersam**
Converts the payload received from the EAB2 frontend into the format required by ePMS.

Leaves Lockup and Storage location hardcoded / removed as they are not yet available. 

The main part of this PR is removeNonEpmsFields.js. This is where the data mapping happens. Where ever possible the mapping configuration has been captured in epmsPayloadConfig.js. This has mostly been the case where fields need renaming or deleting. More complex mappings and things like default values are not in the configuration file. Hopefully this will mean any changes in the future can mostly be configuration rather than code updates. 

The majority of the rest of the changes are example inputs and outputs used for testing. They may be a bit overkill, happy to discuss reducing them if others think so. 

I've based the mapping off a combination of the ePMS API https://epmswebapidev.azurewebsites.net/swagger/index.html and information from the BAs and Product Owners.

### CODE REVIEW

File: client/src/utils/Form/removeNonEpmsFields.js [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+  return copy;
+};
+
+/**
+ * Convert the payload the EAB2 frontend produces into the format required for ePMS
+ * @param {} payload
+ * @returns
+ */
+const getPayloadForEpms = (payload) => {
+  const payloadCopy = payload;
```
Comment (LewisClark-HO) [inline client/src/utils/Form/removeNonEpmsFields.js:154]: I think `payloadCopy` is a confusing name. Any changes made to it are also being applied to the original payload here too. Maybe `payloadReference`?

File: client/src/utils/Form/removeNonEpmsFields.js [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+  fieldsToRemove,
+  itemFieldsToRemove,
+  fieldsToRename,
+  itemFieldsToRename,
+  refDataMappings,
+} from './epmsPayloadConfig';
+
+/**
+ * Applies the mappings found in epmsPayloadConfig to rename/remove/keep fields
+ * @param {*} entry
```
Comment (LewisClark-HO) [inline client/src/utils/Form/removeNonEpmsFields.js:12]: Comments in this file seem a little incomplete, some params are missing descriptions and type specifiers.

File: client/src/utils/Form/removeNonEpmsFields.js [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+ * @param {*} removeArray
+ * @param {*} renameArray
+ * @returns
+ */
+const mapData = (entry, removeArray, renameArray) => {
+  return Object.keys(entry).reduce((acc, key) => {
```
Comment (LewisClark-HO) [inline client/src/utils/Form/removeNonEpmsFields.js:18]: Obviously this is personal preference but I'm not sure what `acc` is meant to be here? I think standard naming for `reduce` is `res` or `result`. A clearer name would be good though I think.

File: client/src/utils/Form/removeNonEpmsFields.js [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+  });
+};
+
+/**
+ * Update the item fields that have a more complex mapping then rename/move/delete
+ * @param {} item the entry object to update
+ * @returns updated item
+ */
+const updateItemFields = (item, obiEmail, meta) => {
+  const copy = item;
```
Comment (LewisClark-HO) [inline client/src/utils/Form/removeNonEpmsFields.js:86]: Again, not sure if it affects implementation but this isn't technically a copy. Any changes made to `copy` here are happening to the original `item` too. Another name might be clearer.

File: client/src/utils/Form/adjustPayloadForType.js [hunk 5]
```diff
@@ -14,17 +15,19 @@ const adjustPayloadForType = (type, formName, payload) => {
...
       submittingUsersEmail: submissionPayload.submittingUsersEmail,
     };
+    submissionPayload = getPayloadForEpms(submissionPayload);
   }
   const jsonPart = {
     type: 'json',
     value: JSON.stringify(submissionPayload),
   };
-  if (type === 'submit' || !id || type === 'submitToEpms') {
+  if (type === 'submit' || !id || type === 'convertForEpms') {
```
Comment (mark-roberts-ho) [inline client/src/utils/Form/adjustPayloadForType.js:30]: If having multiple comparisons to same variable how about:

`[ 'submit'', 'convertForEpms'].includes( type) || !id`

File: client/src/utils/Form/removeNonEpmsFields.js (hunk 3 above)
Comment (mark-roberts-ho) [reply client/src/utils/Form/removeNonEpmsFields.js:18]: `acc` is normally `accumulator`, but a longer name might be good.

File: client/src/utils/Form/removeNonEpmsFields.js (hunk 4 above)
Comment (mark-roberts-ho) [reply client/src/utils/Form/removeNonEpmsFields.js:86]: If it isn't a `copy` there's no need for this constant at all

File: client/src/utils/Form/removeNonEpmsFields.js (hunk 1 above)
Comment (mark-roberts-ho) [reply client/src/utils/Form/removeNonEpmsFields.js:154]: If you're modifying an object, there's no need to have the `reference` or `alias` at all, is there?

**Review by mark-roberts-ho — COMMENTED** [review]
Can you move the exampleXYZ files to a separate sub-directory?

File: client/src/utils/Form/removeNonEpmsFields.js (hunk 1 above)
Comment (gardnersam) [reply client/src/utils/Form/removeNonEpmsFields.js:154]: I did it to satisfy the linter. Directly updating the payload object makes it complain about 
`Assignment to property of function parameter 'payload'  no-param-reassign`
Is there a better way around that?

Comment (gardnersam) [reply client/src/utils/Form/removeNonEpmsFields.js:154]: I could just disable the warning I suppose

File: client/src/utils/Form/removeNonEpmsFields.js (hunk 2 above)
Comment (gardnersam) [reply client/src/utils/Form/removeNonEpmsFields.js:12]: Fair point, I've completed them

File: client/src/utils/Form/removeNonEpmsFields.js (hunk 3 above)
Comment (gardnersam) [reply client/src/utils/Form/removeNonEpmsFields.js:18]: swapped to mappedEntry

File: client/src/utils/Form/removeNonEpmsFields.js (hunk 4 above)
Comment (gardnersam) [reply client/src/utils/Form/removeNonEpmsFields.js:86]: I did this for the same reason, to satisfy the linter, I think I can basically turn off the linter for this check or have the not really needed copy

Comment (LewisClark-HO) [reply client/src/utils/Form/removeNonEpmsFields.js:86]: Seems like a pretty silly check to me, but it must have been turned on for a reason I suppose? My vote would be turn off that check but whatever works.

File: client/src/utils/Form/adjustPayloadForType.js (hunk 5 above)
Comment (gardnersam) [reply client/src/utils/Form/adjustPayloadForType.js:30]: Good shout I've swapped to that

**Review by LewisClark-HO — APPROVED** [review]
Looks good. Thanks for making those changes.

**Pull request merged**